### PR TITLE
[SIM_FLUTTER-105] [BE] Add Category in Withdrawal

### PR DESCRIPTION
**Pull request by IanUrriza10**
### Issue Links
- [backlog](https://framgiaph.backlog.com/view/SIM_FLUTTER-105)
- [slack](https://sun-philippine.slack.com/archives/C059L7XAXFB/p1689062413605599)

### Definition of done
- [x] Make transaction['category'] be able to handle 'FOOD', 'TRANSPORTATION', 'BILLS', 'SAVINGS', 'MISC'
- [x] Return error if there's a mismatch for type and category i.e. DEPT-TRANSPORTATION
- [x] Make a seeder fix class for the changes
- [x] Make a working migration file w/ rollback   

### Screenshots
image
image
image
image

### CODE REVIEW

**Review by Dbulawan13 — DISMISSED** [review]
Approve PR #59 

Working:
- Migration is ok
- Able to update category and add new category options
  -  'FOOD', 'TRANSPORTATION', 'BILLS', 'SAVINGS', 'MISC' 
- Able to withdraw correct amount
- New category options are reflected in the database
- Able to minus current balance
- Unable to withdraw if user input non existing category
- Unable to withdraw if user input insufficient or 0 amount











Testing Hours: 0.58
Comment: 0
Iteration: 0

File: backend/app/Http/Requests/TransactionPostRequest.php [hunk 1]
```diff
@@ -28,10 +28,21 @@ public function rules()
...
+                if ($type == 'DEPT' && ! in_array($value, ['SAVINGS', 'SALARY'])) {
+                    $fail('Wrong transaction category');
+                }
+                if ($type == 'CREDIT' && ! in_array($value, ['FOOD', 'TRANSPORTATION', 'BILLS', 'SAVINGS', 'MISC'])) {
+                    $fail('Wrong transaction category');
+                }
+                if ($type == 'TRANSFER' && ! in_array($value, ['SENDER', 'RECIPIENT'])) {
+                    $fail('Wrong transaction category');
+                }
+            }],
```
Comment (jstephend-sun) [inline backend/app/Http/Requests/TransactionPostRequest.php]: refactor, separate this function

File: backend/database/migrations/2023_07_11_035059_update_transaction_category.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    public function down()
+    {
+
+        Schema::table('transactions', function (Blueprint $table) {
+
+            $transactions = Transaction::where('transaction_type', 'CREDIT')->get();
+            foreach ($transactions as $transaction) {
+                $transaction->category = 'BILLS';
+                $transaction->save();
+            }
```
Comment (jstephend-sun) [inline backend/database/migrations/2023_07_11_035059_update_transaction_category.php]: seems inefficient, use SQL update here, rawsql if necessary, check eloquent first

**Review by jstephend-sun — CHANGES_REQUESTED** [review]
Requested Changes #59

review time: 15mins
comments: 2
iteration: 1st review

File: backend/database/seeders/TransactionSeeder.php [hunk 3]
```diff
@@ -37,11 +37,11 @@ function processAccounts($accounts)
         $type_enums = config('enums.transaction_type');
         $category_enums = config('enums.transaction_category');
 
-        for ($transactionNumber = 0; $transactionNumber <= 100; $transactionNumber++) {
+        for ($transactionNumber = 0; $transactionNumber <= 50; $transactionNumber++) {
             $accounts = Account::inRandomOrder()->take(2)->get();
             $accounts = processAccounts($accounts);
             $transaction_type = $type_enums[rand(0, 2)];
-            $category = $category_enums[2];
+            $category = $category_enums[rand(3, 7)];
```
Comment (jstephend-sun) [inline backend/database/seeders/TransactionSeeder.php:44]: add TODO: improve logic, use subcategories of enum

**Review by jstephend-sun — COMMENTED** [review]
request

**Review by jstephend-sun — APPROVED** [review]
Approved #59

review time: 15mins
comments: 2
iteration: 1st review

**Pull request merged**